src/pip/_internal/vcs/versioncontrol.py

```diff
@@ -422,13 +422,16 @@ def get_url_rev_and_auth(cls, url: str) -> Tuple[str, Optional[str], AuthInfo]:
                     "or remove @ from the URL.".format(url)
                 )
         # check for typos in frag
-        parameter_names = urllib.parse.parse_qs(frag).keys()
-        if not all([p in ("egg", "subdirectory") for p in parameter_names]):
+        unknown_param_names = {
+            key
+            for key in urllib.parse.parse_qs(frag)
+            if key not in ("egg", "subdirectory")
+        }
+        if unknown_param_names:
             raise DiagnosticPipError(
-                reference="test-diagnostic",
-                kind="warning",
-                message="At least one of the URL parameters "
-                "is not supported or has a typo.",
+                reference="typo-in-fragment",
+                message="At least one of the URL parameters {!r}"
+                "is not supported or has a typo.".format(unknown_param_names),
```

Comment:
Would it be friendly to say something like this?

```python
unknown_param_str = ", ".join(sorted(unknown_param_names))
message = f"Not all fragments in the URL {url} are valid: {unknown_param_str}"
```

Reply:
good idea. done.